File: libs/share/share.cpp
#include "pxt.h"
#include "ShareService.h"

enum class ShareLevel{
    all = 0,
    school = 1, 
};


/**
* share
* use
*/

ShareService ShareService(uBit.radio);

//% color=#999999 weight=100 icon="\uf086" block="share"
namespace share {

    int init() {
        return enableRadio();
    }

    int stringToNumber(ManagedString input)
    {
        ManagedString resultString = "";
        int result = 0;

        int l = input.length();
        for(int x=0; x < l; ++x)
        {
            if(input.charAt(x) != '\n' && input.charAt(x) != '\r')
            {
                if(isdigit(input.charAt(x)) || input.charAt(x) == '.')
                {
                    resultString = resultString + input.charAt(x);
                }
                else
                {
                    result = -1;
                    break;
                }
            }
        }
        if(result > -1)
        {
            float f = atof(resultString.toCharArray());
            result = round(f);
        }
        return result;
    }

    //%
    void shareData(StringData* data, StringData* name, ShareLevel level)
    {
        init();
        //This forces the input to be either a 1 or 0. Anything else becomes a 1 (school)
        if(level == ShareLevel::all)
        {
            level = ShareLevel::all;
        }
        else
        {
            level = ShareLevel::school;
        }
        ManagedString result = ShareService.setShareData(ManagedString(data),ManagedString(name),(int)level);
        uBit.display.scroll(result);
    }

    //%
    void shareHistoricalData(StringData* data, StringData* name, StringData* space, StringData* unit)
    {
        init();

        //This forces the input to be either a 1 or 0. Anything else becomes a 1 (school)
        
        ManagedString result = ShareService.setHistoricalData(ManagedString(data),ManagedString(name),ManagedString(space),ManagedString(unit));
        uBit.display.scroll(result);
    }

    //%
    StringData* fetchData(StringData* name, StringData* targetSchool)
    {
        init();

        ManagedString command = "fetchData/";
        command = command + name;
        command = command + "/";
        command = command + targetSchool;
        ManagedString reply = ShareService.getFetchData(command);
        return reply.leakData();
    }

    //%
    StringData* fetchSchoolData(StringData* name)
    {
        init();

        ManagedString command = "fetchData/";
        command = command + name;
        command = command + "/local";
        ManagedString reply = ShareService.getFetchData(command);
        return reply.leakData();
    }

    //%
    int textToNumber(StringData* text)
    {
        init();
        
        int value = stringToNumber((ManagedString)text);
        return value;
    }


    //%
    StringData* numberToText(int value)
    {
        init();
        char buffer[15];
        memset(buffer,0,sizeof(buffer));
        sprintf(buffer,"%d",value);
        return ManagedString(buffer).leakData();
    }

};

File: libs/share/share.ts
//% color=#004d00 weight=100 icon="\uf3fd" block="share"
namespace share {

    //% blockId=share_shareData block="share data %data |called %name |with %level" weight=53
    //% shim=share::shareData
    export function shareData(data:string, name:string, level:ShareLevel): void {
        //basic.showString("OK");
        basic.showIcon(IconNames.Yes);
        return;
    }


    //% blockId=share_shareHistoricalData block="share historical data %data |called %name |namespace %space |units %unit" weight=54
    //% shim=share::shareHistoricalData
    export function shareHistoricalData(data:string, name:string, space:string, unit:string): void {
        //basic.showString("OK");
        basic.showIcon(IconNames.Yes);
        return;
    }

    //% blockId=share_fetchData block="fetch shared data called %name |from school %targetSchool" weight=51
    //% shim=share::fetchData
    export function fetchData(name:string, targetSchool:string): string {
        return "EXAMPLE";
    }

    //% blockId=share_textToNumber block="text to number %text" weight=53
    //% shim=share::textToNumber
    export function textToNumber(text:string): number {
        return parseInt(text);
    }

    //% blockId=share_numberToText block="number to text %value" weight=53
    //% shim=share::numberToText
    export function numberToText(value:number): string {
        return value.toString();
    }

    //% blockId=share_fetchSchoolData block="fetch shared data called %name |from my school" weight=52
    //% shim=share::fetchSchoolData
    export function fetchSchoolData(name:string): string {
        return "EXAMPLE";
    }



    
}